File: android/demos/flipper/src/main/java/com/uber/rib/root/loggedin/tictactoe/TicTacToeInteractor.java
/*
 * Copyright (C) 2017. Uber Technologies
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.uber.rib.root.loggedin.tictactoe;

import androidx.annotation.Nullable;
import com.uber.rib.core.Bundle;
import com.uber.rib.core.Interactor;
import com.uber.rib.core.RibInteractor;
import com.uber.rib.root.loggedin.tictactoe.Board.MarkerType;
import io.reactivex.Observable;
import io.reactivex.functions.Consumer;
import javax.inject.Inject;

/** Coordinates Business Logic for {@link TicTacToeScope}. */
@RibInteractor
public class TicTacToeInteractor
    extends Interactor<TicTacToeInteractor.TicTacToePresenter, TicTacToeRouter> {

  @Inject Board board;
  @Inject TicTacToePresenter presenter;

  private final String playerOne = "Fake name 1";
  private final String playerTwo = "Fake name 2";

  private MarkerType currentPlayer = MarkerType.CROSS;

  @Override
  protected void didBecomeActive(@Nullable Bundle savedInstanceState) {
    super.didBecomeActive(savedInstanceState);

    presenter
        .squareClicks()
        .subscribe(
            new Consumer<BoardCoordinate>() {
              @Override
              public void accept(BoardCoordinate xy) throws Exception {
                if (board.cells[xy.getX()][xy.getY()] == null) {
                  if (currentPlayer == MarkerType.CROSS) {
                    board.cells[xy.getX()][xy.getY()] = MarkerType.CROSS;
                    board.currentRow = xy.getX();
                    board.currentCol = xy.getY();
                    presenter.addCross(xy);
                    currentPlayer = MarkerType.NOUGHT;
                  } else {
                    board.cells[xy.getX()][xy.getY()] = MarkerType.NOUGHT;
                    board.currentRow = xy.getX();
                    board.currentCol = xy.getY();
                    presenter.addNought(xy);
                    currentPlayer = MarkerType.CROSS;
                  }
                }
                if (board.hasWon(MarkerType.CROSS)) {
                  presenter.setPlayerWon(playerOne);
                } else if (board.hasWon(MarkerType.NOUGHT)) {
                  presenter.setPlayerWon(playerTwo);
                } else if (board.isDraw()) {
                  presenter.setPlayerTie();
                } else {
                  updateCurrentPlayer();
                }
              }
            });
    updateCurrentPlayer();
  }

  private void updateCurrentPlayer() {
    if (currentPlayer == MarkerType.CROSS) {
      presenter.setCurrentPlayerName(playerOne);
    } else {
      presenter.setCurrentPlayerName(playerTwo);
    }
  }

  /** Presenter interface implemented by this RIB's view. */
  interface TicTacToePresenter {
    Observable<BoardCoordinate> squareClicks();

    void setCurrentPlayerName(String currentPlayer);

    void setPlayerWon(String playerName);

    void setPlayerTie();

    void addCross(BoardCoordinate xy);

    void addNought(BoardCoordinate xy);
  }

  public interface Listener {}
}

File: android/demos/compose/src/main/kotlin/com/uber/rib/compose/root/main/logged_in/tic_tac_toe/TicTacToeInteractor.kt
/*
 * Copyright (C) 2021. Uber Technologies
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.uber.rib.compose.root.main.logged_in.tic_tac_toe

import com.uber.rib.compose.root.main.AuthInfo
import com.uber.rib.compose.util.EventStream
import com.uber.rib.compose.util.StateStream
import com.uber.rib.core.BasicInteractor
import com.uber.rib.core.Bundle
import com.uber.rib.core.ComposePresenter
import com.uber.rib.core.coroutineScope
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach

class TicTacToeInteractor(
  presenter: ComposePresenter,
  private val authInfo: AuthInfo,
  private val eventStream: EventStream<TicTacToeEvent>,
  private val stateStream: StateStream<TicTacToeViewModel>,
  private val listener: Listener
) : BasicInteractor<ComposePresenter, TicTacToeRouter>(presenter) {

  var currentPlayer: Board.MarkerType = Board.MarkerType.CROSS

  override fun didBecomeActive(savedInstanceState: Bundle?) {
    super.didBecomeActive(savedInstanceState)

    eventStream.observe()
      .onEach {
        when (it) {
          is TicTacToeEvent.BoardClick -> {
            val board: Board = stateStream.current().board
            val coord = it.coordinate

            if (board.cells[coord.x][coord.y] == null) {
              if (currentPlayer == Board.MarkerType.CROSS) {
                board.cells[coord.x][coord.y] = Board.MarkerType.CROSS
                board.currentRow = coord.x
                board.currentCol = coord.y
                currentPlayer = Board.MarkerType.NOUGHT
              } else {
                board.cells[coord.x][coord.y] = Board.MarkerType.NOUGHT
                board.currentRow = coord.x
                board.currentCol = coord.y
                currentPlayer = Board.MarkerType.CROSS
              }
            }

            if (board.hasWon(Board.MarkerType.CROSS)) {
              listener.onGameWon(authInfo.playerOne)
            } else if (board.hasWon(Board.MarkerType.NOUGHT)) {
              listener.onGameWon(authInfo.playerTwo)
            } else if (board.isDraw()) {
              listener.onGameWon(null)
            }

            val newPlayerName = if (currentPlayer == Board.MarkerType.CROSS) {
              authInfo.playerOne
            } else {
              authInfo.playerTwo
            }

            stateStream.dispatch(
              stateStream.current().copy(
                board = board,
                currentPlayer = newPlayerName
              )
            )
          }
          TicTacToeEvent.XpButtonClick -> TODO("Go somewhere")
        }
      }.launchIn(coroutineScope)
  }

  interface Listener {
    fun onGameWon(winnerName: String?)
  }
}
